README.md
# RedSage

Maintain sanity while combining Redmine activity times and Sage project times. This is the deal of RedSage: Sum up correct data into false output data for sanity's sake.

The output times are totally off in favor of easier data input in Sage, that is: every new time slot starts at a full hour. That may mean for a lot of project switches and one official break in a day, Sage-wise will report you to _officially_ work from 08:00 till 22:00 o'clock while in reality you worked from 07:45 till 16:45.

1. Run Redmine query
1. Save .CSV
1. Run RedSage
1. Enter formatted console output in Sage

```bash
# tell RedSage to have a lunch break of 60 minutes (default)
./redsage redmine.csv

# tell RedSage to have a lunch break of 45 minutes 
./redsage -b 45 redmine.csv

# show help topics
./redsage --help
./redsage run --help
```

German CSV quickstart:

Calling RedSage like this reads a Redmine time report .CSV in german locale and joins all pipelines into a single one. Currently, lunch break default to 12:00 o'clock with a duration of 60 minutes.

```
redsage run -s "Gesamtzeit" -c ";" -d "," -i /path/to/timelog-1.csv

Pipeline A    7,50    6,00    4,50    4,50    22,50   
Pipeline B    1,50                            1,50    
ACME                          0,75    0,75    
Pipeline A-joined
2021-05-03      08:00 - 12:00   13:00 - 18:00   
2021-05-04      08:00 - 12:00   13:00 - 15:00   
2021-05-05      08:00 - 12:00   13:00 - 13:30   
2021-05-06      08:00 - 12:00   13:00 - 14:15   
```

Used .CSV:
```csv
Anforderungspipeline;2021-05-03;2021-05-04;2021-05-05;2021-05-06;Gesamtzeit
Pipeline A;7,50;6,00;4,50;4,50;22,50
Pipeline B;1,50;"";"";"";1,50
ACME;"";"";"";0,75;0,75
Gesamtzeit;9,00;6,00;4,50;5,25;24,75
```

## License

MIT

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added
- initial release

cmd/run_test.go
package cmd

import (
	"testing"
)

func TestRun(t *testing.T) {

}


redsage_test.go
package main

import (
	"bufio"
	"github.com/stretchr/testify/require"
	"io/ioutil"
	"os"
	"testing"
)

func Test_doRun(t *testing.T) {
	t.Run("should join 2 pipelines and break them up with lunch breaks", func(t *testing.T) {
		file, _ := ioutil.TempFile(os.TempDir(), "redmineCSV-")
		path := file.Name()
		defer os.Remove(path)
		csvWriter := bufio.NewWriter(file)
		_, _ = csvWriter.Write([]byte(`Anforderungspipeline;2021-05-03;2021-05-04;2021-05-05;2021-05-06;Gesamtzeit
Pipeline A;7,50;6,00;"";4,50;18,00
Pipeline B/2;1,50;2,00;"";3,50;6,00
Gesamtzeit;7,50;6,00;"";4,50;18,00
`))
		args := runArgs{
			lunchBreakInMin:  60,
			singlePipelines:  []string{},
			filename:         path,
			csvDelimiter:     ";",
			decimalDelimiter: ",",
			skipColumnNames:  []string{"Gesamtzeit"},
			skipSummaryLine:  true,
		}

		// when
		actual := doRun(args)

		// then
		require.NoError(t, actual)
	})
}


core/util_test.go
package core

import (
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"testing"
	"time"
)

func TestParseDateWithTime(t *testing.T) {
	t.Run("should parse time", func(t *testing.T) {
		actual, err := ParseDateWithTime("2021-05-05", "17:00:00")

		require.NoError(t, err)
		expected := time.Date(2021, 05, 05, 17, 0, 0, 0, time.UTC)
		assert.Equal(t, expected, actual)
	})
	t.Run("should error for missing seconds", func(t *testing.T) {
		_, err := ParseDateWithTime("2021-05-05", "17:00")

		require.Error(t, err)
	})
}


redsage.go
package main

import (
	"fmt"
	"github.com/pkg/errors"
	"github.com/ppxl/sagemine/core"
	"github.com/ppxl/sagemine/cruncher"
	"github.com/ppxl/sagemine/logging"
	"github.com/ppxl/sagemine/reader"
	"github.com/ppxl/sagemine/transformer"
	"github.com/sirupsen/logrus"
	"github.com/urfave/cli/v2"
	"os"
)

const (
	flagGlobalLogLevel           = "log-level"
	flagLunchBreakInMinutesLong  = "break"
	flagLunchBreakInMinutesShort = "b"
	flagSinglePipelinesLong      = "pipeline-single"
	flagSinglePipelinesShort     = "p"
	flagCSVColumnDelimiterLong   = "csv-column-delimiter"
	flagCSVColumnDelimiterShort  = "c"
	flagDecimalDelimiterLong     = "decimal-delimiter"
	flagDecimalDelimiterShort    = "d"
	flagIgnoreSummaryLineLong    = "ignore-summary-line"
	flagIgnoreSummaryLineShort   = "i"
	flagSkipColumnsLong          = "skip-column"
	flagSkipColumnsShort         = "s"
)

var (
	// Version of the application
	Version string
	log     = logrus.New()
)

type runArgs struct {
	lunchBreakInMin  int
	singlePipelines  []string
	filename         string
	csvDelimiter     string
	decimalDelimiter string
	skipColumnNames  []string
	skipSummaryLine  bool
}

func createGlobalFlags() []cli.Flag {
	return []cli.Flag{
		&cli.StringFlag{
			Name:  flagGlobalLogLevel,
			Usage: "define log level",
			Value: "warning",
		},
	}
}

func configureLogging(cliCtx *cli.Context) error {
	logLevel := cliCtx.String("log-level")
	logLevelParsed, err := logrus.ParseLevel(logLevel)
	if err != nil {
		log.Errorf("could not parse log level %s to logrus level. Defaulting to WARN", logLevel)
		logLevelParsed = logrus.WarnLevel
	}

	err = logging.Init(logLevelParsed)
	log = logging.Logger()
	if err != nil {
		return errors.Wrap(err, "could not initialize logging")
	}
	return nil
}

func configureApplication(cliCtx *cli.Context) error {
	err := configureLogging(cliCtx)
	if err != nil {
		return err
	}

	return nil
}

func checkMainError(err error) {
	if err != nil {
		println("%+s\n", err)
		os.Exit(1)
	}
}

// projects main function
func main() {
	app := cli.NewApp()
	app.Name = "redsage"
	app.Usage = "Maintain sanity while combining Redmine activity times and Sage project times"
	app.Version = Version
	app.Commands = []*cli.Command{run()}

	app.Flags = createGlobalFlags()
	app.Before = configureApplication
	err := app.Run(os.Args)
	checkMainError(err)
}

func run() *cli.Command {
	return &cli.Command{
		Name:      "run",
		Usage:     "read Redmine work time data and convert them to Sage-compatible data",
		Action:    doCliRun,
		ArgsUsage: "redmine CSV file",
		Flags: []cli.Flag{
			&cli.IntFlag{
				Name:    flagLunchBreakInMinutesLong,
				Aliases: []string{flagLunchBreakInMinutesShort},
				Usage:   "lunch break time in minutes (optional)",
				Value:   60,
			},
			&cli.StringSliceFlag{
				Name:    flagSinglePipelinesLong,
				Aliases: []string{flagSinglePipelinesShort},
				Usage: "These pipelines will receive their own pipeline and will not be joint into a single pseudo-pipeline (optional). " +
					"All other pipelines will be merged into a single pseudo-pipeline.",
			},
			&cli.StringSliceFlag{
				Name:    flagSkipColumnsLong,
				Aliases: []string{flagSkipColumnsShort},
				Usage:   "columns with these headers will be ignored (optional)",
			},
			&cli.StringFlag{
				Name:    flagCSVColumnDelimiterLong,
				Aliases: []string{flagCSVColumnDelimiterShort},
				Usage:   "this delimiter will be used to parse CSV columns (optional)",
				Value:   ";",
			},
			&cli.StringFlag{
				Name:    flagDecimalDelimiterLong,
				Aliases: []string{flagDecimalDelimiterShort},
				Usage:   "Set the decimal delimiter if the decimals in the CSV export uses a different format than '2.75' (optional)",
				Value:   ".",
			},
			&cli.BoolFlag{
				Name:    flagIgnoreSummaryLineLong,
				Aliases: []string{flagIgnoreSummaryLineShort},
				Usage:   "Set if the last line in the CSV export should be included or nto (optional)",
				Value:   true,
			},
		},
	}
}

func doCliRun(cliCtx *cli.Context) error {
	if cliCtx.Args().Len() < 1 {
		_ = cli.ShowAppHelp(cliCtx)
		return errors.New("filename argument missed")
	}
	if cliCtx.Args().Len() > 1 {
		_ = cli.ShowAppHelp(cliCtx)
		return fmt.Errorf("found more arguments than expected: '%v'", cliCtx.Args().Slice()[1:])
	}

	filename := cliCtx.Args().First()
	lunchBreakInMin := cliCtx.Int(flagLunchBreakInMinutesLong)
	csvColumnDelimiter := cliCtx.String(flagCSVColumnDelimiterLong)
	decimalDelimiter := cliCtx.String(flagDecimalDelimiterLong)
	ignoreSummaryLine := cliCtx.Bool(flagIgnoreSummaryLineLong)
	singlePipelines := cliCtx.StringSlice(flagSinglePipelinesLong)
	skipColumns := cliCtx.StringSlice(flagSkipColumnsLong)

	args := runArgs{
		lunchBreakInMin:  lunchBreakInMin,
		singlePipelines:  singlePipelines,
		filename:         filename,
		csvDelimiter:     csvColumnDelimiter,
		decimalDelimiter: decimalDelimiter,
		skipColumnNames:  skipColumns,
		skipSummaryLine:  ignoreSummaryLine,
	}

	return doRun(args)
}

func doRun(args runArgs) error {
	data, err := readRedmineData(args)
	if err != nil {
		return err
	}

	joinedData, err := joinRedmineData(data, args)
	if err != nil {
		return err
	}

	crunched, err := crunch(joinedData, args)
	if err != nil {
		return err
	}

	printResults(crunched)

	return nil
}

func printResults(crunched *core.CrunchedOutput) {
	for i, pipeline := range crunched.NamedDaySageValues {
		fmt.Printf("%v\n", i)
		for _, date := range pipeline.SortedKeys() {
			fmt.Printf("%s\t", date)
			for _, timeslot := range pipeline.TimeSlots(date) {
				fmt.Printf("%s\t", timeslot.String())
			}
			fmt.Println()
		}
	}
}

func joinRedmineData(data *core.PipelineData, args runArgs) (*core.PipelineData, error) {
	trans := transformer.New()
	joinConfig := transformer.Config{
		SinglePipelineNames: args.singlePipelines,
	}

	joinedData, err := trans.Transform(data, joinConfig)
	if err != nil {
		return nil, errors.Wrap(err, "error while transforming pipelines")
	}

	return joinedData, nil
}

func readRedmineData(args runArgs) (*core.PipelineData, error) {
	options := reader.Options{
		Type: reader.CSV,
		CSVOptions: reader.CSVOptions{
			Filename:         args.filename,
			CSVDelimiter:     args.csvDelimiter,
			DecimalDelimiter: args.decimalDelimiter,
			SkipColumnNames:  args.skipColumnNames,
			SkipSummaryLine:  args.skipSummaryLine,
		},
		APIOptions: reader.APIOptions{},
	}
	redmineReader := reader.New(options)

	data, err := redmineReader.Read()
	if err != nil {
		return nil, errors.Wrapf(err, "error while reading from %s", options.CSVOptions.Filename)
	}

	return data, nil
}

func crunch(data *core.PipelineData, args runArgs) (*core.CrunchedOutput, error) {
	crunchConfig := cruncher.Config{
		LunchBreakInMin: args.lunchBreakInMin,
	}
	crunch := cruncher.New()

	crunched, err := crunch.Crunch(data, crunchConfig)
	if err != nil {
		return nil, errors.Wrapf(err, "error while crunching data")
	}

	return crunched, nil
}


reader/csvReader.go
package reader

import (
	"encoding/csv"
	"fmt"
	"github.com/pkg/errors"
	"github.com/ppxl/sagemine/core"
	"github.com/ppxl/sagemine/logging"
	"os"
	"strconv"
	"strings"
)

const (
	CSV = iota
	RestAPI
)

var log = logging.Logger()

type CSVOptions struct {
	Filename         string
	CSVDelimiter     string
	DecimalDelimiter string
	SkipColumnNames  []string
	SkipSummaryLine  bool
}

type APIOptions struct {
	RedmineURL      string
	RedmineUser     string
	RedminePassword string
}

type Options struct {
	Type       int
	CSVOptions CSVOptions
	APIOptions APIOptions
}

type RedmineDataReader interface {
	Read() (*core.PipelineData, error)
}

type csvReader struct {
	options CSVOptions
}

func New(options Options) *csvReader {
	switch options.Type {
	case CSV:
		return newCSVReader(options.CSVOptions)
	case RestAPI:
		fallthrough
	default:
		log.Panicf("unsupported Redmine reader type %d", options.Type)
	}
	return nil
}

func newCSVReader(options CSVOptions) *csvReader {
	return &csvReader{options: options}
}

func (cr *csvReader) Read() (*core.PipelineData, error) {
	fileReader, err := os.OpenFile(cr.options.Filename, os.O_RDWR|os.O_CREATE, os.ModePerm)
	if err != nil {
		panic(err)
	}
	defer fileReader.Close()
	commaRunes := []rune(cr.options.CSVDelimiter)

	r := csv.NewReader(fileReader)
	r.Comma = commaRunes[0]
	r.Comment = '#'

	data, err := r.ReadAll()
	if err != nil {
		return nil, err
	}

	result := core.NewPipelineData()
	columnHeaders := []string{}
	columnsToSkip := []int{}

	for currentLine, line := range data {
		var pipeline *core.RedmineWorkPerDay

		if currentLine == 0 {
			columnHeaders = line
			columnsToSkip = buildSkipColumns(columnHeaders, cr.options.SkipColumnNames)
			continue
		}

		if cr.options.SkipSummaryLine && isLastLine(currentLine, data) {
			break
		}

		for currentColumn, cell := range line {
			fmt.Printf("%s\t", cell)

			if currentColumn == 0 {
				pipeline, err = result.AddPipeline(cell)
				if err != nil {
					return nil, errors.Wrapf(err, "failed to read line %d from CSV: error while adding pipeline %s", currentLine, cell)
				}
				continue
			}

			if skipColumn(currentColumn, columnsToSkip) {
				continue
			}

			workTimeRaw := formatDecimal(cell, cr.options.DecimalDelimiter)

			workTime, err := strconv.ParseFloat(workTimeRaw, 64)
			if err != nil {
				return nil, errors.Wrapf(err, "could not cast value '%s' to float (line %d, column %d)", cell, currentLine, currentColumn)
			}

			currentDay := columnHeaders[currentColumn]

			pipeline.PutWorkTime(currentDay, workTime)
		}

		fmt.Println()
	}

	return result, nil
}

func isLastLine(currentLine int, data [][]string) bool {
	return currentLine == len(data)-1
}

func skipColumn(currentColumn int, skip []int) bool {
	for _, skipColumn := range skip {
		if currentColumn == skipColumn {
			return true
		}
	}

	return false
}

func buildSkipColumns(headers []string, headerToBeSkipped []string) []int {
	result := []int{}
	for index, header := range headers {
		for _, skip := range headerToBeSkipped {
			if header == skip {
				result = append(result, index)
			}
		}
	}

	return result
}

func formatDecimal(cell string, delimiter string) string {
	if cell == "" {
		return "0.0"
	}

	floatDecimalDelimiter := "."
	return strings.Replace(cell, delimiter, floatDecimalDelimiter, 1)
}


cruncher/cruncher.go
package cruncher

import (
	"github.com/pkg/errors"
	"github.com/ppxl/sagemine/core"
	"github.com/sirupsen/logrus"
	"time"
)

const (
	dayStartTime    = "08:00:00"
	lunchStartTime  = "12:00:00"
	wallClockLayout = "15:04"
)

// Config contains configuration values that modify the number crunching behaviour.
type Config struct {
	LunchBreakInMin int
}

// Cruncher provides methods for transforming values from a redmine pipeline data.
type Cruncher interface {
	// Crunch executes the merging and splitting values from a redmine pipeline data and returns them.
	Crunch(pdata *core.PipelineData, config Config) (*core.CrunchedOutput, error)
}

type cruncher struct {
}

func New() *cruncher {
	return &cruncher{}
}

const shouldStartNextSlotAtFullOClock = false

// Crunch executes the merging and splitting values from a CSV file and prints the output in Sage-relatable manner.
func (c *cruncher) Crunch(pdata *core.PipelineData, config Config) (*core.CrunchedOutput, error) {
	output := core.NewCrunchedOutput()
	dayTimeCounter := core.NewDayTimeCounter(dayStartTime)

	for redminePipeline, workPerDay := range pdata.NamedDayRedmineValues {
		logrus.Debugf("Add new pipeline %s", redminePipeline)
		pipelineName := string(redminePipeline)

		pipeline, err := output.AddPipeline(pipelineName)
		if err != nil {
			return nil, errors.Wrap(err, "error while crunching time data")
		}

		for day, worktime := range workPerDay.WorkPerDay {
			currentDayAndTime := dayTimeCounter.GetNextTimeSlotOrDefault(day)

			if containsNoWorkTime(worktime) {
				pipeline.PutEmptyTimeSlot(day)
				continue
			}

			start := currentDayAndTime.Format(wallClockLayout)

			// decimals don't work well with duration: Do instead manual minute calculation
			calcedEndTime := time.Duration(worktime*60) * time.Minute
			endTime := currentDayAndTime.Add(calcedEndTime)

			diff, endTimeIntersectsWithLunchtime := endTimeFallsIntoLunch(endTime, day)
			if endTimeIntersectsWithLunchtime {
				logrus.Debugf("Time slot %s - %s falls into lunch by %s. Breaking up into two parts...", currentDayAndTime, endTime, diff)
				endBeforeLunch := endTime.Add(-diff)
				pipeline.PutTimeSlot(day, start, endBeforeLunch.Format(wallClockLayout))
				// update current time to enable correct timing of the second slot
				currentDayAndTime = endBeforeLunch

				startAfterLunch := currentDayAndTime.Add(time.Duration(config.LunchBreakInMin) * time.Minute)
				endAfterLunch := startAfterLunch.Add(diff)
				pipeline.PutTimeSlot(day, startAfterLunch.Format(wallClockLayout), endAfterLunch.Format(wallClockLayout))
				currentDayAndTime = endAfterLunch
			} else {
				end := endTime.Format(wallClockLayout)
				pipeline.PutTimeSlot(day, start, end)
				currentDayAndTime = endTime
			}

			dayTimeCounter.EndTime(day, currentDayAndTime)
		}
	}

	return output, nil
}

// endTimeFallsIntoLunch returns false if the given time does not overlap with to configured lunch time. Otherwise true
// and the (aboslute) duration of the overlap will be returned.
func endTimeFallsIntoLunch(workTimeEnd time.Time, day string) (time.Duration, bool) {
	parsedLunchtime, err := time.Parse(time.RFC3339, day+"T"+lunchStartTime+"Z")
	if err != nil {
		panic("lunchtime: " + err.Error())
	}

	fallsIntoLunchTime := workTimeEnd.After(parsedLunchtime)
	if fallsIntoLunchTime {
		return workTimeEnd.Sub(parsedLunchtime), true
	}

	return 0, false
}

func roundWorkTimeToNextHour(currentWorkTime time.Time) time.Time {
	// go one hour forward and subtract the actual minutes to get the full next hour
	roundedNextHour := currentWorkTime.Add(1 * time.Hour).Add(-time.Duration(currentWorkTime.Minute()) * time.Minute)
	return roundedNextHour
}

func startsWorkAtOClock(workTime time.Time) bool {
	return workTime.Minute() == 0
}

func containsNoWorkTime(worktime float64) bool {
	return worktime == 0.0
}


transformer/joinPipeline.go
package transformer

import (
	"github.com/pkg/errors"
	"github.com/ppxl/sagemine/core"
)

// Config contains configuration values that modify the joining behaviour.
type Config struct {
	SinglePipelineNames []string
}

type Transformer interface {
	Transform(pdata *core.PipelineData, config Config) (*core.PipelineData, error)
}

func New() *joinTransformer {
	return &joinTransformer{}
}

type joinTransformer struct {
}

func (j *joinTransformer) Transform(pdata *core.PipelineData, config Config) (*core.PipelineData, error) {
	result := core.NewPipelineData()
	var err error
	var joinedPipeline *core.RedmineWorkPerDay

	for redminePipeline, workPerDay := range pdata.NamedDayRedmineValues {
		if joinedPipeline == nil {
			pipelineName := string(redminePipeline) + "-joined"
			joinedPipeline, err = result.AddPipeline(pipelineName)

			if err != nil {
				return nil, errors.Wrap(err, "error while crunching time data")
			}
		}

		for date, worktime := range workPerDay.WorkPerDay {
			joinedPipeline.PutWorkTime(date, worktime)
		}
	}

	return result, nil
}


core/types.go
package core

import (
	"errors"
	"fmt"
	"github.com/sirupsen/logrus"
	"sort"
	"time"
)

const timeSlotFormat = "%s - %s"

func NewPipelineData() *PipelineData {
	values := make(map[PipelineName]*RedmineWorkPerDay, 0)
	return &PipelineData{NamedDayRedmineValues: values}
}

// PipelineData contain parsed Redmine Pipeline data. Redmine divides hour in a decimal way, i. e. 0.5 means 30 minutes.
//
// Example:
//  data := NewPipelineData()
//  pipeline, err := data.AddPipeline("My Pipeline")
//  pipeline.PutWorkTime("2021-05-05", 7.5)
//  pipeline.PutWorkTime("2021-05-06", 4)
type PipelineData struct {
	// NamedDayValues maps the pipeline name to actual values per day, f. i. Pipeline 1 -> 2021-05-05 -> 5.75
	NamedDayRedmineValues map[PipelineName]*RedmineWorkPerDay
}

func (pd *PipelineData) Entries() int {
	return len(pd.NamedDayRedmineValues)
}

func (pd *PipelineData) AddPipeline(pipelineName string) (*RedmineWorkPerDay, error) {
	if pipelineName == "" {
		return nil, errors.New("pipeline name must not be empty")
	}

	pipeline := newRedmineWorkPerDay()
	pd.NamedDayRedmineValues[(PipelineName)(pipelineName)] = pipeline

	return pipeline, nil
}

// RedmineWorkPerDay maps a date string to the accumulated amount of time spent, f. i. 2021-05-05 -> 5.75
type RedmineWorkPerDay struct {
	WorkPerDay map[string]float64
}

func newRedmineWorkPerDay() *RedmineWorkPerDay {
	values := make(map[string]float64, 0)
	return &RedmineWorkPerDay{WorkPerDay: values}
}

func (rwpd *RedmineWorkPerDay) Days() int {
	return len(rwpd.WorkPerDay)
}

func (rwpd *RedmineWorkPerDay) PutWorkTime(date string, workTime float64) {
	currentWorkTime := rwpd.WorkTime(date)
	currentWorkTime += workTime

	rwpd.WorkPerDay[date] = currentWorkTime
}

func (rwpd *RedmineWorkPerDay) WorkTime(date string) float64 {
	return rwpd.WorkPerDay[date]
}

// CrunchedOutput contains mappings from pipeline name to Sage compatible work time
type CrunchedOutput struct {
	NamedDaySageValues map[PipelineName]*SageWorkPerDay
}

func (co *CrunchedOutput) String() string {
	result := ""

	sortedKeys := co.SortedKeys()
	for _, dayValue := range sortedKeys {
		logrus.Info("String: dayValue: " + dayValue)
		result += fmt.Sprintf("p: %s", co.NamedDaySageValues[(PipelineName)(dayValue)])
	}

	return result
}

func (co *CrunchedOutput) SortedKeys() []string {
	keys := make([]string, 0, len(co.NamedDaySageValues))
	for k := range co.NamedDaySageValues {
		keys = append(keys, (string)(k))
	}
	sort.Strings(keys)

	return keys
}

func NewCrunchedOutput() *CrunchedOutput {
	values := make(map[PipelineName]*SageWorkPerDay, 0)
	return &CrunchedOutput{NamedDaySageValues: values}
}

func (co *CrunchedOutput) AddPipeline(pipelineName string) (*SageWorkPerDay, error) {
	if pipelineName == "" {
		return nil, errors.New("pipeline name must not be empty")
	}

	pipeline := &SageWorkPerDay{}
	co.NamedDaySageValues[(PipelineName)(pipelineName)] = pipeline

	return pipeline, nil
}

// SageWorkPerDay maps a date string to a simplified Sage time slow, f. i. 2021-05-05 -> 13:00 - 14:00
type SageWorkPerDay map[string][]TimeSlot

func (swpd *SageWorkPerDay) Days() int {
	return len(*swpd)
}

func (swpd *SageWorkPerDay) PutTimeSlot(date string, slotStart, slotEnd string) {
	timeSlots := swpd.TimeSlots(date)

	timeSlot := TimeSlot{}
	timeSlot.Start = slotStart
	timeSlot.End = slotEnd

	timeSlots = append(timeSlots, timeSlot)

	(*swpd)[date] = timeSlots
}

func (swpd *SageWorkPerDay) String() string {
	result := ""
	sortedKeys := swpd.SortedKeys()
	for _, timeSlot := range sortedKeys {
		result += fmt.Sprintf("d: ts: %s\t", timeSlot)
	}

	return result
}

func (swpd *SageWorkPerDay) SortedKeys() []string {
	keys := make([]string, 0, len(*swpd))
	for k := range *swpd {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	return keys
}

func (swpd *SageWorkPerDay) TimeSlots(day string) []TimeSlot {
	return (*swpd)[day]
}

func (swpd *SageWorkPerDay) PutEmptyTimeSlot(day string) {
	swpd.PutTimeSlot(day, "-", "-")
}

// TimeSlot represents a dateless wall clock interval of work, f. i. from 13:00 till 14:15
type TimeSlot struct {
	// Start contains the time slot's ending time in 24-hour format, f. i. "13:00" for 1 pm
	Start string
	// End contains the time slot's ending time in 24-hour format, f. i. "13:00" for 1 pm
	End string
}

func (t *TimeSlot) String() string {
	return fmt.Sprintf(timeSlotFormat, t.Start, t.End)
}

// PipelineName contains the name of a pipeline.
type PipelineName string

// DayTimeCounter holds the state of timeslots per day regardless of project.
type DayTimeCounter struct {
	counters             map[string]time.Time
	defaultWorkStartTime string
}

func NewDayTimeCounter(workStartTime string) *DayTimeCounter {
	counters := make(map[string]time.Time, 0)
	return &DayTimeCounter{counters: counters, defaultWorkStartTime: workStartTime}
}

func (dtc *DayTimeCounter) GetNextTimeSlotOrDefault(date string) time.Time {
	result, ok := dtc.counters[date]
	if !ok {
		goodMorning, err := ParseDateWithTime(date, dtc.defaultWorkStartTime)
		if err != nil {
			panic("could not get end time for day: " + err.Error())
		}
		dtc.counters[date] = goodMorning
		return goodMorning
	}

	return result
}
func (dtc *DayTimeCounter) EndTime(date string, endTime time.Time) {
	dtc.counters[date] = endTime
}
